# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1463166119 14400
#      Fri May 13 15:01:59 2016 -0400
# Node ID 2cb0dbd5154f37e90d99c33d4e9dba94b90ff554
# Parent  de138ebaa7a946f76712730f5ae16875b1607b5b
# Parent  2c6d1ef071d1104ddcc46ce403adc4b919bf9619
maint: Merge stable to default.

diff --git a/libinterp/dldfcn/audioread.cc b/libinterp/dldfcn/audioread.cc
--- a/libinterp/dldfcn/audioread.cc
+++ b/libinterp/dldfcn/audioread.cc
@@ -339,47 +339,54 @@ Comment.\n\
   std::string title = "";
   std::string artist = "";
   std::string comment = "";
   // Quality is currently unused?
   //
   // float quality = 0.75;
   for (int i = 3; i < nargin; i += 2)
     {
-      if (args(i).string_value () == "BitsPerSample")
+      if (i >= nargin - 1)
+        error ("audiowrite: invalid number of arguments");
+
+      std::string keyword = args(i).string_value ();
+      octave_value value_arg = args(i+1);
+
+      if (keyword == "BitsPerSample")
         {
           info.format &= ~SF_FORMAT_SUBMASK;
-          int bits = args(i + 1).int_value ();
+          int bits = value_arg.int_value ();
           if (bits == 8)
             {
               if ((info.format & SF_FORMAT_TYPEMASK) == SF_FORMAT_WAV)
                 info.format |= SF_FORMAT_PCM_U8;
               else
                 info.format |= SF_FORMAT_PCM_S8;
             }
           else if (bits == 16)
             info.format |= SF_FORMAT_PCM_16;
           else if (bits == 24)
             info.format |= SF_FORMAT_PCM_24;
           else if (bits == 32)
             info.format |= SF_FORMAT_PCM_32;
           else
             error ("audiowrite: wrong number of bits specified");
         }
-      else if (args(i).string_value () == "BitRate")
+      else if (keyword == "BitRate")
         ;
       // Quality is currently unused?
-      // else if (args(i).string_value () == "Quality")
-      //   quality = args(i + 1).int_value () * 0.01;
-      else if (args(i).string_value () == "Title")
-        title = args(i + 1).string_value ();
-      else if (args(i).string_value () == "Artist")
-        artist = args(i + 1).string_value ();
-      else if (args(i).string_value () == "Comment")
-        comment = args(i + 1).string_value ();
+      //
+      // else if (keyword == "Quality")
+      //   quality = value_arg.int_value () * 0.01;
+      else if (keyword == "Title")
+        title = value_arg.string_value ();
+      else if (keyword == "Artist")
+        artist = value_arg.string_value ();
+      else if (keyword == "Comment")
+        comment = value_arg.string_value ();
       else
         error ("audiowrite: wrong argument name");
     }
 
   SNDFILE *file = sf_open (filename.c_str (), SFM_WRITE, &info);
 
   if (! file)
     error ("audiowrite: failed to open output file %s", filename.c_str ());
